File: cpp/Algo/games/task2/task2.cpp
#include <iostream>
#include <vector>
#include <set>
#include <map>
#include <algorithm>

const int n = 8;

struct Position {
    int r = 0, t = 0;
    bool isRunnerStep = false;

    Position(int r=0, int t=0, bool s=true) : r(r), t(t), isRunnerStep(s) {}

    bool operator<(const Position& p) const {
        if (r < p.r)
            return true;
        if (r > p.r)
            return false;
        if (t < p.t) 
            return true;
        if (t > p.t)
            return false;
        if (!isRunnerStep && p.isRunnerStep)
            return true;
        if (isRunnerStep && p.isRunnerStep)
            return false;
        return false;
        
    }

    bool operator==(const Position& p) const {
        return r == p.r ? (t == p.t ? (isRunnerStep == p.isRunnerStep) : false) : false;
    } 
};

std::ostream & operator<<(std::ostream &out, const Position& pos) {
    out << "{" << pos.r << ", " << pos.t << "} " << pos.isRunnerStep;
    return out;
}

class Game {
    std::map<Position, std::vector<Position>> g;
    Position startPosition;
    std::set<Position> usedPositions;
    std::set<Position> winPositions;
    std::set<Position> losePositions;

    std::set<std::pair<int, int>> walls;
public:
    Game(int runner, int termy, std::set<std::pair<int, int>> &walls) : 
        startPosition(runner, termy, true),
        walls(walls) {        
    }
private:

    bool isRunnerStep(Position &pos) {
        return pos.isRunnerStep;
    }

    bool isTermyWin(Position &pos) {
        int x1 = pos.r % n;
        int y1 = pos.r / n;

        int x2 = pos.t % n;
        int y2 = pos.t / n;

        if (x1 == x2 && y1 == y2) {
            return true;
        }

        if (x1 - x2 == 0 || y1 - y2 == 0 || abs(x1 - x2) == abs(y1 - y2)) {
            // check walls
            if (x1 == x2) {
                for (int y = std::min(y1, y2) + 1; y < std::max(y1, y2); y++) {
                    if (walls.count({y, x1}) != 0) {
                        return false;
                    }
                }
            } else if (y1 == y2) {
                for (int x = std::min(x1, x2) + 1; x < std::max(x1, x2); x++) {
                    if (walls.count({y1, x}) != 0) {
                        return false;
                    }
                }
            } else {
                int stepx = x2 > x1 ? 1 : -1;
                int stepy = y2 > y1 ? 1 : -1;
                for (int x = x1 + stepx, y = y1 + stepy; 
                    x < x2; x += stepx, y += stepy) {
                    if (walls.count({y, x}) != 0) {
                        return false;
                    }
                }
            }
            return true;
        }
        
        return false;
    }

    bool isRunnerWin(Position &pos) {
        if (isTermyWin(pos)) {
            return false;
        }

        if (isRunnerStep(pos)) {
            int y1 = pos.r / n;

            if (y1 == n-1)
                return true;
        }
            
        return false;
    }

    bool isTerminal(Position &pos) {
        if (isRunnerWin(pos))
            return true;
                
        if (isTermyWin(pos))
            return true;
        
        return false;
    }

    std::vector<Position> makeStep(Position &pos) {
        const int dx[] = {-1, -1, -1,  0, 0,  1, 1, 1};
        const int dy[] = {-1,  0,  1, -1, 1, -1, 0, 1};

        int x, y, px, py, runner, termy;

        std::vector<Position> nextPositions;

        if (isRunnerStep(pos)) {
            x = pos.r % n;
            y = pos.r / n;
        } else {
            x = pos.t % n;
            y = pos.t / n;
        }

        for (int i = 0; i < 8; i++) {
            px = x + dx[i];
            py = y + dy[i];

            if (px < 0 || px >= n || py < 0 || py >= n) continue;

            if (walls.count({y, x}) != 0) {
                continue;
            }

            if (isRunnerStep(pos)) {
                runner = py * n + px;
                termy = pos.t;
            } else {
                runner = pos.r;
                termy = py * n + px;
            }

            bool step = !pos.isRunnerStep;

            nextPositions.push_back({runner, termy, step});
        }

        usedPositions.insert(pos);

        return nextPositions;
    }

public:
int solve() {
        std::vector<Position> q;
       
        q.push_back(startPosition);

        while (q.size() > 0) {
            Position position = q[0];
            std::cerr << "Get from queue: " << q[0] << std::endl;
            q.erase(q.begin());

            int count1 = usedPositions.size();
            /*
            usedPositions.insert(position);
            if (position == Position({20, 17, true})) {
                std::cerr << "Handle " << position << " | usedPositions.count(position) = " << usedPositions.count(position) << std::endl;
                std::cerr << "Handle " << position << " | count(position) = " << std::count(usedPositions.begin(), usedPositions.end(), position) << std::endl;

                for (auto &pos : usedPositions) {
                    if (pos == Position({20, 17, true})) {
                        std::cerr << "Found" << std::endl;
                        break;
                    }
                }

                throw 2;
            }
            */
            
            if (isTerminal(position)) {
                if (isRunnerStep(position)) {
                    if (isRunnerWin(position)) {
                        //std::cerr << "Runner Win - add " << position << " to N" << std::endl;
                        winPositions.insert(position);
                    } else {
                        //std::cerr << "Runner Lose - add " << position << " to P" << std::endl;
                        losePositions.insert(position);
                    }
                } else {
                    if (isTermyWin(position)) {
                        //std::cerr << "Termy Win - add " << position << " to N" << std::endl;
                        winPositions.insert(position);
                    } else {
                        //std::cerr << "Termy Lose - add " << position << " to P" << std::endl;
                        losePositions.insert(position);
                    }
                }
                usedPositions.insert(position);
            } else {
                g[position] = makeStep(position);
                
                for (auto &nextPosition : g[position]) {
                    
                    if (usedPositions.count(nextPosition) == 0 &&
                        std::find(q.begin(), q.end(), nextPosition) == q.end()) {
                        //std::cerr << nextPosition << " - Add to queue" << std::endl;
                        q.push_back(nextPosition);
                    }
                    
                }
                
            }
            int count2 = usedPositions.size();
            std::cerr << "Pos count: " << count2 << " | Q count: " << q.size() << std::endl;
            if (count1 == count2) {
                throw 2;
                break;
            }
        }

        std::set<Position> &N = winPositions;
        std::set<Position> &P = losePositions;
        std::set<Position> U = usedPositions;


        std::set<Position> diff;
        std::set<Position> un;

        std::set_union(N.begin(), N.end(), P.begin(), P.end(), std::inserter(un, un.begin()));
        std::set_difference(U.begin(), U.end(), un.begin(), un.end(), std::inserter(diff, diff.begin()));
        U = diff;

        //print('Start position:', startPosition)
        //print(f'Start: U: {len(U)} | N: {len(N)} | P: {len(P)}')
        
        while (U.size() > 0) {
            std::set<Position> R;
            for (auto &u : U) {
                int c = 0;
                for (auto &v : g[u]) {
                    if (P.count(v) > 0) {
                        // print(u, '- Add to N')
                        N.insert(u); 
                        R.insert(u);
                        break;
                    }
                    if (N.count(v) > 0) {
                        c += 1;
                    }
                } 
                if (c == g[u].size()) {
                    // print(u, '- Add to P')
                    P.insert(u);
                    R.insert(u);
                }
            }
            if (R.size() > 0) {
                diff.clear();
                std::set_difference(U.begin(), U.end(), R.begin(), R.end(), std::inserter(diff, diff.begin()));
                U = diff;
            } else {
                // print(f'Break U: {len(U)} | N: {len(N)} | P: {len(P)}') 
                break;
            }

            // print(f'U: {len(U)} | N: {len(N)} | P: {len(P)}')          

        }

        if (N.count(startPosition) > 0) {
            return 1;
        } 
        return -1;        
    }
};



int main() {
    freopen("input.txt", "r", stdin);
    
    int runner = 0;
    int termy = 0;
    std::set<std::pair<int, int>> walls;

    for (int i = 0; i < n; i++) {
        std::string s;
        std::cin >> s;
        // std::cerr << s << std::endl;
        for (int j = 0; j < n; j++) {
            if (s[j] == '1') {
                walls.insert({i, j});
                // std::cerr << "Wall at " << i << ", " << j << std::endl;
            } else if (s[j] == '2') {
                runner = i*n + j;
            } else if (s[j] == '3') {
                termy = i*n + j;
            }
        }
    }

    Game game(runner, termy, walls);

    int ans = game.solve();

    std::cout << ans << std::endl;


    return 0;
}

File: cpp/Algo/games/task2/task2.py
# https://www.e-olymp.com/ru/submissions/8012686
import sys
sys.setrecursionlimit(10000)
n = 8

usedPositions = set()
winPositions = set()
losePositions = set()

walls = set()

g = dict()

def isRunnerStep(pos) -> bool:
    return pos[2]

def isTermyWin(pos, walls) -> bool:
    x1 = pos[0] % n
    y1 = pos[0] // n

    x2 = pos[1] % n
    y2 = pos[1] // n

    if x1 == x2 and y1 == y2:
        return True

    if (x1 - x2 == 0 or y1 - y2 == 0 or abs(x1 - x2) == abs(y1 - y2)):
        # check walls
        if x1 == x2:
            for y in range(min(y1, y2) + 1, max(y1, y2)):
                if (y, x1) in walls:
                    return False
        elif y1 == y2:
            for x in range(min(x1, x2) + 1, max(x1, x2)):
                if (y1, x) in walls:
                    return False
        else:
            stepx = 1 if x2 > x1 else -1
            stepy = 1 if y2 > y1 else -1
            for x, y in zip(range(x1 + stepx, x2, stepx), range(y1 + stepy, y2, stepy)):
                if (y, x) in walls:
                    return False
               
        return True
    
    return False

def isRunnerWin(pos, walls) -> bool:
    if isTermyWin(pos, walls): 
        return False

    if isRunnerStep(pos):
        y1 = pos[0] // n

        if y1 == n-1:
            return True
        
    return False


def isTerminal(pos, walls) -> bool:
    if isRunnerWin(pos, walls):
        return True
            
    if isTermyWin(pos, walls):
        return True
    
    return False


def makeStep(pos, walls): 
    dx = [-1, -1, -1,  0, 0,  1, 1, 1]
    dy = [-1,  0,  1, -1, 1, -1, 0, 1]

    nextPositions = []

    if isRunnerStep(pos):
        x = pos[0] % n
        y = pos[0] // n
    else:
        x = pos[1] % n
        y = pos[1] // n
    
    for i in range(8):
        px = x + dx[i]
        py = y + dy[i]

        if px < 0 or px >= n or py < 0 or py >= n:
            continue

        if (py, px) in walls:
            continue
        
        if isRunnerStep(pos):
            runner = py * n + px
            termy = pos[1]
        else:
            runner = pos[0]
            termy = py * n + px

        step = not pos[2]

        nextPositions.append((runner, termy, step))
    
    usedPositions.add(pos)
    
    return nextPositions
'''
used = set()
def dfs(u, N, P):
    used.add(u)
    print(f'{u} - Used: {len(used)} | N: {len(N)} | P: {len(P)}')
    if u in N or u in P:
        return
    for v in g[u]:
        if v not in used:
            dfs(v, N, P)
    c = 0
    for v in g[u]:
        if v in P:
            N.add(u)
            break
        if v in N:
            c += 1
    else:
        if (c == len(g[u])):
            P.add(u)
        else:
            pass
            # raise ValueError
'''

def solve(runner, termy, walls):
    startPosition = (runner, termy, True)

    q = []

    q.append(startPosition)

    while len(q) > 0:
        position = q[0]
        # print("Get from queue:", position)
        del q[0]

        if isTerminal(position, walls):
            if isRunnerStep(position):
                if isRunnerWin(position, walls):
                    # print(f'Runner Win - add {position} to N')
                    winPositions.add(position)
                else:
                    # print(f'Runner Lose - add {position} to P')
                    losePositions.add(position)
            else:
                if isTermyWin(position, walls):
                    # print(f'Termy Win - add {position} to N')
                    winPositions.add(position)
                else:
                    # print(f'Termy Lose - add {position} to P')
                    losePositions.add(position)

            usedPositions.add(position)
        else:
            nextPositions = makeStep(position, walls)
            g[position] = nextPositions
            for nextPosition in nextPositions:
                if nextPosition not in usedPositions and \
                    nextPosition not in q:
                    # print(nextPosition, '- Add to queue')
                    q.append(nextPosition)

        # print(f'Pos count: {len(usedPositions)} | Q count: {len(q)}')

    # print(len(winPositions))
    # print(len(losePositions))
    # print(len(usedPositions))

    N = winPositions
    P = losePositions

    

    U = usedPositions
    U = (U - P) - N

    '''
    for position in P | N:
        if position not in used:
            dfs(position, N, P)
            print(f'U: {len(U)} | N: {len(N)} | P: {len(P)}')
    '''
    # print('Start position:', startPosition)
    print(f'Start: U: {len(U)} | N: {len(N)} | P: {len(P)}')
    
    while len(U) > 0 and startPosition not in N and startPosition not in P:
        R = set()
        for u in U:
            c = 0
            for v in g[u]:
                if v in P:
                    # print(u, '- Add to N')
                    N.add(u) 
                    R.add(u)
                    break
                if v in N:
                    c += 1
            else:
                if c == len(g[u]):
                    # print(u, '- Add to P')
                    P.add(u)
                    R.add(u)
        if len(R) > 0:
            U -= R
        else:
            # print(f'Break U: {len(U)} | N: {len(N)} | P: {len(P)}') 
            break

        # print(f'U: {len(U)} | N: {len(N)} | P: {len(P)}')          


    if startPosition in winPositions:
        print(1)
    elif startPosition in losePositions:
        print(-1)
    else:
        print(-2)


f = open('input.txt', 'r')

for i in range(n):
    s = f.readline().strip()
    for j in range(n):
        if s[j] == '1':
            walls.add((i, j))
        elif s[j] == '2':
            runner = i*n + j
        elif s[j] == '3':
            termy = i*n + j

f.close()

solve(runner, termy, walls)